### PR TITLE
Enhance manifests upgrade

### PR DESCRIPTION
**Pull request by johnliu55tw**
**IMPORTANT: Please do not create a Pull Request without creating an issue first.**

**Problem:**
While upgrading manifests (rancher and harvester and their dependencies), we need to ensure each component is updated before we move on to upgrading the next component.

**Solution:**
Use different strategies to wait for different components.

**Related Issue:**
- https://github.com/harvester/harvester/issues/1982

**Dependencies**
- https://github.com/harvester/harvester-installer/pull/252

**Test plan:**
1. Build an ISO from harvester-installer using PR https://github.com/harvester/harvester-installer/pull/252
2. Build `harvester-upgrade` container image using this PR
3. Test upgrade using the ISO and the `harvester-upgrade` image

### CODE REVIEW

**Comment by bk201** [comment]
NOTE this PR depends on the installer change in https://github.com/harvester/harvester-installer/pull/252

File: package/upgrade/upgrade_manifests.sh [hunk 1]
```diff
@@ -38,19 +142,44 @@ upgrade_rancher()
...
   until [ "$(get_running_rancher_version)" = "$REPO_RANCHER_VERSION" ]
   do
     echo "Wait for Rancher to be upgraded..."
     sleep 5
   done
+
+  echo "Wait for Rancher dependencies helm releases..."
+  wait_helm_release cattle-fleet-system fleet $REPO_FLEET_CHART_VERSION $REPO_FLEET_APP_VERSION deployed
+  wait_helm_release cattle-fleet-system fleet-crd $REPO_FLEET_CRD_CHART_VERSION $REPO_FLEET_CRD_APP_VERSION deployed
+  wait_helm_release cattle-system rancher-webhook $REPO_RANCHER_WEBHOOK_CHART_VERSION $REPO_RANCHER_WEBHOOK_APP_VERSION deployed
```
Comment (bk201) [inline package/upgrade/upgrade_manifests.sh]:
```suggestion
  wait_helm_release cattle-fleet-system fleet fleet-$REPO_FLEET_CHART_VERSION $REPO_FLEET_APP_VERSION deployed
  wait_helm_release cattle-fleet-system fleet-crd fleet-crd-$REPO_FLEET_CRD_CHART_VERSION $REPO_FLEET_CRD_APP_VERSION deployed
  wait_helm_release cattle-system rancher-webhook rancher-webhook-$REPO_RANCHER_WEBHOOK_CHART_VERSION $REPO_RANCHER_WEBHOOK_APP_VERSION deployed
```

https://github.com/harvester/harvester-installer/pull/252 output versions in harvester-release.yaml without chart name, would be nice to prefix the chart name here.

File: package/upgrade/upgrade_manifests.sh [hunk 2]
```diff
@@ -94,9 +245,16 @@ EOF
   upgrade_managed_chart_from_version $UPGRADE_PREVIOUS_VERSION harvester harvester.yaml
   NEW_VERSION=$REPO_HARVESTER_CHART_VERSION yq e '.spec.version = strenv(NEW_VERSION)' harvester.yaml -i
   kubectl apply -f ./harvester.yaml
+
+  pause_managed_chart harvester "false"
+  pause_managed_chart harvester-crd "false"
+
```
Comment (bk201) [inline package/upgrade/upgrade_manifests.sh:287]: Would be nice to wait managed chart to ready here: please see https://github.com/bk201/harvester/commit/314ab334d80a7cb9a7417e52bd4c1a9339c403fb#diff-8c5fdd3c29c4031d6332031bafd4f1c97820a492be51384b2b22689c6322d31fR267-R268

File: package/upgrade/upgrade_manifests.sh [hunk 3]
```diff
@@ -111,10 +269,15 @@ spec:
   version: $REPO_MONITORING_CHART_VERSION
 EOF
   kubectl patch managedcharts.management.cattle.io rancher-monitoring -n fleet-local --patch-file ./rancher-monitoring.yaml --type merge
+
+  pause_managed_chart rancher-monitoring "false"
+  pause_managed_chart rancher-monitoring-crd "false"
```
Comment (bk201) [inline package/upgrade/upgrade_manifests.sh:324]: dittos, to wait managed chart to be ready: https://github.com/bk201/harvester/commit/314ab334d80a7cb9a7417e52bd4c1a9339c403fb#diff-8c5fdd3c29c4031d6332031bafd4f1c97820a492be51384b2b22689c6322d31fR295-R296

Comment (johnliu55tw) [reply package/upgrade/upgrade_manifests.sh:324]: Totally missed that part. Thanks for the reminder!

**Review by bk201 — APPROVED** [review]
LGTM.

File: package/upgrade/upgrade_manifests.sh [hunk 4]
```diff
@@ -5,13 +5,150 @@ UPGRADE_TMP_DIR="/tmp/upgrade"
...
+      if [ "$current_observed_generation" -gt "$generation" ]; then
+        summary_state=$(echo "$current_chart" | yq e ".status.summary.$state" -)
+        if [ "$summary_state" = "1" ]; then
+          break
+        fi
+      fi
+    fi
+
+    sleep 5
+    echo "Sleep for 5 seconds to retry"
```
Comment (weihanglo) [inline package/upgrade/upgrade_manifests.sh:36]: Do we need a timeout abort here Or leaving it to the job deadline is sufficient?

**Review by weihanglo — COMMENTED** [review]
Have you tested single-node upgrade path? I cannot succeed on single-node cluster and there seems no memory or storage insufficient event.
It always errored up during **Images preloading** stage and then kube-vip, kubevirt, and other components restarted. After that I got an Error from upgrade-repo VM and never come back.

File: package/upgrade/upgrade_manifests.sh (hunk 4 above)
Comment (bk201) [reply package/upgrade/upgrade_manifests.sh:36]: job deadline should be good.

**Comment by bk201** [comment]
@weihanglo Probably hit [node pressure eviction](https://kubernetes.io/docs/concepts/scheduling-eviction/node-pressure-eviction/#hard-eviction-thresholds). Try keeping 25G free space before the upgrade.

**Review by weihanglo — APPROVED** [review]
I cannot let it pass on my machine but there was lots of entanglement. The code itself looks good to me. I'll approve it and let's see how far it can go.

**Pull request merged**